### ISSUE TITLE
Minimum window size is larger than screen size (on Windows)

### ISSUE THREAD

**Issue by nicost** [napari/napari#4290]
Installed napari using conda on Windows 10 laptop.  Run napari using external monitor (higher resolution then build-in screen). When now running napari on laptop screen, it is impossible to size the window so that it fits on the screen (other than by maximizing the windows altogether).  The minimum window size seems to be set to something larger than the actual display.  In addition, while resizing the window, it regularly disappears from the display altogether, and the only way to get it back is to close the application from the task manager and start it up again.  Output in the anaconda terminal:  
 
```
C:\miniconda3\envs\napari-env\lib\site-packages\napari\_qt\__init__.py:53: UserWarning:

napari was tested with QT library `>=5.12.3`.
The version installed is 5.9.7. Please report any issues with
this specific QT version at https://github.com/Napari/napari/issues.
  warn(message=warn_message)
WARNING: QWindowsWindow::setGeometry: Unable to set geometry 1204x1148+12+46 on QWidgetWindow/'_QtMainWindowClassWindow'. Resulting geometry:  1204x1076+12+46 (frame: 11, 45, 11, 11, custom margin: 0, 0, 0, 0, minimum size: 602x538, maximum size: 16777215x16777215).
09:22:28 WARNING QWindowsWindow::setGeometry: Unable to set geometry 1204x1148+12+46 on QWidgetWindow/'_QtMainWindowClassWindow'. Resulting geometry:  1204x1076+12+46 (frame: 11, 45, 11, 11, custom margin: 0, 0, 0, 0, minimum size: 602x538, maximum size: 16777215x16777215).
```

Napari Info:

napari: 0.4.14
Platform: Windows-10-10.0.19041-SP0
Python: 3.9.10 | packaged by conda-forge | (main, Feb 1 2022, 21:22:07) [MSC v.1929 64 bit (AMD64)]
Qt: 5.9.7
PyQt5: 5.9.2
NumPy: 1.21.5
SciPy: 1.7.3
Dask: 2022.02.1
VisPy: 0.9.6

OpenGL:
- GL version: 4.6.0 - Build 30.0.100.9955
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 960x540, scale 2.0

Plugins:
- console: 0.0.4
- scikit-image: 0.4.14
- svg: 0.1.5

**Comment by tlambert03** [comment]
Thanks @nicost, I do think we've probably overlooked things as small as 540px, so indeed we should go through and double check any assumptions.  (@napari/core-devs, the primary window resizing spot is [right here](https://github.com/napari/napari/blob/main/napari/_qt/qt_main_window.py#L1027), and it looks to be trying to work with the sizeHint of the layout... presumably bigger than 540px).

I think this is also partially an issue with [set_window_settings](https://github.com/napari/napari/blob/main/napari/_qt/qt_main_window.py#L210) restoring geometry from a previous session (i.e. one with a larger monitor) in a new session that can't accommodate the settings.  This should  fail gracefully.

**Comment by tlambert03** [comment]
sidenote @nicost. This isn't your issue here, but that Qt warning you're getting about `The version installed is 5.9.7.` suggests that you have a rather old version of Qt in your environment:  This is very easy to do if one installs things that use pyqt from the "defaults" anaconda channel (your default if you simply installed miniconda and didn't modify your channels).

I would recommend creating a new environment and either installing everything from conda forge:
```
conda create -n napari-env -c conda-forge napari
```

or from pip:
```
conda create -n napari-env -c conda-forge python=3.9
conda activate napari-env
pip install "napari[all]"
```

**Comment by nicost** [comment]
Laptop display size is 1920x1080, so is something scaling by a factor of 2?  Can I delete settings to test without history of a previous display?

**Comment by tlambert03** [comment]
`napari --reset`!

**Comment by nicost** [comment]
Nice!  Same issue though;(

**Comment by tlambert03** [comment]
yeah, we've also definitely struggled with handling windows monitor scaling nicely in all cases.  I guess you have scaling set to 2 in your general Windows "Display" preferences?

**Comment by nicost** [comment]
It is set to 150%(recommended).  Display sure looks better at 100% or 125% (in both those cases, the napari window fits easily)!  Not sure how 150% scaling of 1080 pixels results in 540  pixels.

**Comment by nicost** [comment]
B.t.w., I installed from conda-forge using the instructions on https://napari.org/tutorials/fundamentals/installation.  So I guess that conda-forge is still at qt 5.9.7?

**Comment by tlambert03** [comment]
yeah, very confusing 🤔  ... I admit, display scaling and the handling thereof is rather mysterious to me.  We need someone to generally dig into this.  I'm marking this with some priority and severity... since it comes up a lot.  cc @sofroniewn: we need a general windows scaling and Qt hero here :joy:

> B.t.w., I installed from conda-forge using the instructions on https://napari.org/tutorials/fundamentals/installation. So I guess that conda-forge is still at qt 5.9.7?

conda-forge is shipping 5.12.3: https://anaconda.org/conda-forge/pyqt ... my guess is that you installed into a pre-existing environment instead of a new one?  (if the installer sees pyqt is already there in your environment, it's not going to overwrite it with the conda forge version)

**Comment by nicost** [comment]
>  conda-forge is shipping 5.12.3: https://anaconda.org/conda-forge/pyqt ... my guess is that you installed into a pre-existing environment instead of a new one? (if the installer sees pyqt is already there in your environment, it's not going to overwrite it with the conda forge version)

I following the conda installation instructions to the letter.  Repeated it to be sure.  Noticed the following after `conda update napari`:

 ```
qt                      conda-forge::qt-5.12.9-h5909a2a_4 --> pkgs/main::qt-5.9.7-vc14h73c81de_0
 qtpy                 conda-forge::qtpy-2.0.1-pyhd8ed1ab_0 --> pkgs/main::qtpy-1.11.2-pyhd3eb1b0_0
```

Python package management is a continuing source of misery for me;(.  I guess that some dependency insists on using qt 5.9.7?

**Comment by tlambert03** [comment]
you're absolutely right.  I just tried it.
So sorry for the confusion (and it _is_ terribly confusing).  The docs need additional clarification, as they essentially assume that people have the conda-forge channel, and in some cases `channel_priority=strict`, added to their conda config.

I do have an explanation for you, but no one should be expected to know this. It's a real mess :/
here's my summary advice (explanation below).  Do either of the following:

1. **use conda _only_ for environment management, and only use pip for package management**
   ```sh
   # creates a new conda environment, and installs python 3.9 from the conda forge channel
   conda create --name my-new-env --channel conda-forge python=3.9
   # activate new environment
   conda activate my-new-env
   # install napari from pip with qt backend
   pip install "napari[all]"
   # an should you want to update napari:
   pip install -U napari
   ```
2. If you'd like to install from conda rather than pip, **add conda-forge to your conda config, and set `channel_priority` to `strict`**:
   ```sh
   conda config --add channels conda-forge
   conda config --set channel_priority strict
   # then you can create a new environment and install/update napari without specifying channel
   conda create --name my-new-env napari
   ```

-------

and here's the long explanation (elaborating a bit on the whole conda, conda-forge, pip thing) if you're curious:

while conda is an excellent environment manager, with great ability to solve environments for deterministic builds, the real "weak link" in the chain here (particularly when it comes to Qt based applications) is the ["defaults" channel](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/channels.html).  Unlike pip, when you install from conda, you can install a package from one of many different package channels.  You can do this either by using the `-c channel-name` flag when installing, or by adding a channel permanently to your config.  The "defaults" channel (the only one in the default configuration when you install conda from anaconda or miniconda) is actually rather lacking.  Almost everyone who uses conda a lot eventually adds the [conda forge](https://conda-forge.org/) channel to their configuration.  **AND** if you follow conda-forge's installation advice, you'll also set your ["channel priority" to strict](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html?highlight=channel_priority#strict-channel-priority).  Conda-forge is the saving grace of conda environments :) it should absolutely be _the_ default channel (and, if you install [miniforge](https://github.com/conda-forge/miniforge) instead of anaconda/miniconda, it will be!)

```
conda config --add channels conda-forge
conda config --set channel_priority strict
```

When we write installation instructions, such as on the napari docs, we tend to add `-c conda-forge` to the install command as a "safe guard" just in case the user hadn't added conda forge to their config.  It turns out though, as you found, that there _are_ cases, depending on your channel priority, and general conda config, that you can still get pyqt from the defaults channel (which remains at 5.9) even if you use `-c conda-forge`.  I had to tweak my conda config to see this, but I definitely saw it too.

So, if I were you, I would
1. create new environments liberally – all the time
2. make sure to add the conda forge channel to your config
3. set channel priority to strict
4. in the case of napari, try just installing from pip instead of conda. 
5. if you'd like, you can try out the bundled applications, added as assets to each release on github: https://github.com/napari/napari/releases

**Comment by tlambert03** [comment]
oh btw, one more thing:  I gave a seminar trying to demystify this all just a little bit, and it's recorded and available here if you're interested: https://nic.med.harvard.edu/python/

**Comment by sofroniewn** [comment]
Just catching up on this one - 

> The docs need additional clarification, as they essentially assume that people have the conda-forge channel, and in some cases channel_priority=strict, added to their conda config.

Should we add this to the docs? Maybe @DragaDoncila can make a PR based on https://github.com/napari/napari/issues/4290#issuecomment-1072916223. Also apologies if this has already been done, I'm super behind on notifications, catching up now

**Comment by cnstt** [comment]
For a visualisation purpose, I am building a plugin that is handling 3 viewers at the same time on the screen, and I would like to be able to set the window size (height and width a little bit smaller than the default minimum values, to be able to see them all at the same time.
So I am also interested in better tweaking options of the min window size, which might be related to this issue :)

**Comment by tlambert03** [comment]
took a quick look at this.  Looks like there is no longer any minimum size imposed on the MainWindow object itself, but it will depend on all other widgets inside.  So if _anything_ sets a minimumSize, the main window will expand to contain it.

I think we should aim for the viewer itself (just canvas without extra widgets) to be as flexible as possible.  Currently, if you create a viewer and hide the layer controls and the layer list, the main window still can't get smaller than ~(618, 563).  Looks like this is largely due to size limitations imposed by the `QtWelcomeWidget` (if you hide it, the window can get much smaller).  So making that widget more flexible with regards to sizing would be a good start. 

cc @goanpeca

**Comment by cnstt** [comment]
Thanks a lot for your explanation!

> Looks like this is largely due to size limitations imposed by the `QtWelcomeWidget` (if you hide it, the window can get much smaller).


What is this `QtWelcomeWidget` and how could I hide it?

**Comment by goanpeca** [comment]
Thanks for the ping @tlambert03!

@cnstt this is the widget in question https://github.com/napari/napari/blob/5f96d5d814aad697c367bdadbb1a57750e2114ad/napari/_qt/widgets/qt_welcome.py
<img width="1151" alt="Screen Shot 2022-04-27 at 11 06 35 AM" src="https://user-images.githubusercontent.com/3627835/165562707-87513f82-0604-4ec2-89c0-bd768b0c6d93.png">

The image is the one imposing a size restriction.
https://github.com/napari/napari/blob/5f96d5d814aad697c367bdadbb1a57750e2114ad/napari/_qt/widgets/qt_welcome.py#L37

**Comment by cnstt** [comment]
> @cnstt this is the widget in question https://github.com/napari/napari/blob/5f96d5d814aad697c367bdadbb1a57750e2114ad/napari/_qt/widgets/qt_welcome.py

Thanks for your help! I tried to hide this welcome widget, but the viewer is still blocked to the same min height. What could this come from? From the image layer I have added afterwards (can they trigger a min height?) ?

**Comment by goanpeca** [comment]
The height is blocked by the controls dock and the layers dock. I also commented the image size line and it can go as small as the actual text in the welcome widget allows. We could enforce it going below it, but still the control and layer docks are imposing some minimum size constraints.

<img width="441" alt="Screen Shot 2022-04-27 at 12 04 27 PM" src="https://user-images.githubusercontent.com/3627835/165580723-ef2c3531-6d35-4359-b53d-1fe77d06e9ce.png">

**Comment by jni** [comment]
> Not sure how 150% scaling of 1080 pixels results in 540 pixels.

This seems to be a Windows scaling/Qt thing: anything at or above 150 is taken to be 2x. You can see @GenevieveBuckley's screenshots of napari at various scaling values [in #1710](https://github.com/napari/napari/issues/1710#issuecomment-712489853). The UI elements only have 1x or 2x size.
